### PR TITLE
Fix service clients initial fail

### PR DESCRIPTION
**Pull request by weichou1229**

### CODE REVIEW

File: utils_test.go [hunk 1]
```diff
@@ -66,13 +66,13 @@ func TestConsulClientReturnErrorOnBadResponse(t *testing.T) {
...
 
-	consul, err := getConsulClient(config)
+	consul, err := GetConsulClient("test-service-name", "test-profile", config)
 	if consul != nil || err == nil {
 		t.Error("Error should be raised")
 	}
 
-	if err.Error() != "Bad response from Consul service" {
-		t.Error("Wrong error message")
+	if err.Error() != "Cannot get connection to consul" {
```
Comment (falkerson) [inline utils_test.go]: why this is changed? We have separate test for this called 'TestConsulClientReturnErrorOnTimeout'

**Comment by falkerson** [comment]
@weichou1229 could you please describe in commit what an issue current commit fix? What is exactly the issue?

**Comment by weichou1229** [comment]
@falkerson It's https://github.com/edgexfoundry/device-sdk-go/issues/47

File: clientinitializer.go [hunk 2]
```diff
@@ -195,36 +202,53 @@ func initializeClients() {
...
+	svc.ec = coredata.NewEventClient(params, consulEndpoint)
 
 	params.Path = v1Valuedescriptor
 	params.Url = dataAddr + params.Path
-	svc.vdc = coredata.NewValueDescriptorClient(params, types.Endpoint{})
+	svc.vdc = coredata.NewValueDescriptorClient(params, consulEndpoint)
+
+}
+
+func checkClientsInitialSuccessful() {
```
Comment (falkerson) [inline clientinitializer.go]: this one should checks successful initialization of all clients or only schedules and scheduleEvents? From function name description - all, but in code I see only two. Maybe better add comments for function name or change it accordingly?

File: examples/simple/cmd/res/configuration.toml [hunk 3]
```diff
@@ -2,7 +2,7 @@
 Host = "localhost"
 Port = 49990
 ConnectRetries = 3
-HealthCheck = ""
+HealthCheck = "/api/v1/ping"
```
Comment (falkerson) [inline examples/simple/cmd/res/configuration.toml]: pls make it in separate commit. This one related to test simple-device service in examples not for device-sdk

File: registry/consul.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"fmt"
+	consulapi "github.com/hashicorp/consul/api"
+)
+
+type ConsulClient struct {
```
Comment (falkerson) [inline registry/consul.go]: what is purpose of this structure?

File: utils_test.go (hunk 1 above)
Comment (weichou1229) [reply utils_test.go]: I have rewrote the consul functions, so I made some change for test.

File: clientinitializer.go (hunk 2 above)
Comment (weichou1229) [reply clientinitializer.go]: ok, I will add comments on it.

**Comment by falkerson** [comment]
@weichou1229 yes I understand but git history should be self descriptive. It's about maintenance. For example some people take this repo for own needs and they don't understand anything whats going on from changelog which based on git log history not github issues. Moreover they could have their own ALM like Gitlab with own git server. In simple words git history and github issues are not the same.

File: utils_test.go (hunk 1 above)
Comment (falkerson) [reply utils_test.go]: case with bad response still in code https://github.com/edgexfoundry/device-sdk-go/pull/50/files#diff-7d1c2a3334601b6c1958aae0a594cba8R130 and we have no test for it. Moreover currently we have two tests which check same case with 'Cannot get connection to consul'.

File: examples/simple/cmd/res/configuration.toml (hunk 3 above)
Comment (falkerson) [reply examples/simple/cmd/res/configuration.toml]: yes, sure. But 'simple' directory is not related to device-sdk. It even can be in separate repository. So pull request should contain 2 separate commits. One for device-sdk and another onw for simple-device example.

File: registry/consul.go (hunk 4 above)
Comment (falkerson) [reply registry/consul.go]: Sorry maybe I've asked wrong, but I mean 'ConsulClient struct'. Currently it's simple proxy with 1 field. What sense of such structure? Do you want make some additional logic in it? If not - currently it complicates things.

File: examples/simple/cmd/res/configuration.toml (hunk 3 above)
Comment (weichou1229) [reply examples/simple/cmd/res/configuration.toml]: Sorry, I got your point. It's need more commit history to describe changes.

File: registry/consul.go (hunk 4 above)
Comment (weichou1229) [reply registry/consul.go]: These are copy from edgex internal package.
And I made a interface for consul.
It contains three functions: 1.init 2.registry 3.load consul config
https://github.com/edgexfoundry/device-sdk-go/blob/53a1df7d6021b66f9b7fd1d3468e6c2b0634e688/registry/registry.go

File: utils.go [hunk 5]
```diff
@@ -156,6 +77,65 @@ func LoadConfig(profile string, configDir string) (config *Config, err error) {
 	return config, nil
 }
 
+func GetConsulClient(serviceName string, profile string, config *Config) (*registry.ConsulClient, error) {
```
Comment (falkerson) [inline utils.go]: this one not just returns consul client but also does many things with registry. IMHO: logic is too complicated. We have registryClient, consulClient and a bit inconsistency between function names and logic in it. As for me we should:

1. Check if we could connect to consul.
2. If yes: return consul client else return proper error message
3. register service and health check with consul client
4. check key value pairs
5. get config from consul

File: registry/consul.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		if key == serviceKey {
+			endpoint.Port = service.Port
+			endpoint.Key = key
+			endpoint.Address = service.Address
+		}
+	}
+	return endpoint, nil
+}
+
+func (c *ConsulClient) CheckKeyValuePairs(configuration interface{}, applicationName string, profiles []string) error {
```
Comment (falkerson) [inline registry/consul.go]: did you check it? if this one copy/paste from edgex-go I'm afraid it will not work properly with device configuration structure.

In edgex-go configs have flat structure like:
checkUrl ="bla"
port = 8000
serviceName = ""

in device-sdk we had groups like:
[Service]
  Name = "bla"

So algorithm for parsing from edgex-go doesn't work properly for device-sdk

File: utils.go (hunk 5 above)
Comment (weichou1229) [reply utils.go]: Actually I want to continue integrate this package until we can direct access this package.
https://github.com/edgexfoundry/edgex-go/tree/master/internal/pkg/consul
Then we can do refactor according this https://github.com/edgexfoundry/edgex-go/blob/master/internal/core/command/init.go#L100

Comment (falkerson) [reply utils.go]: you can. IMHO we need to keep things granular: if function called GetConsulClient it should returns only consul client and doesn't do any other changes implicitly, for such case we can do next:

in serviceStart:
1. get consul client
2. register service with consul client if it's not
3. register health with consul client check if it's not
4. check key-value with consul client

File: registry/consul.go (hunk 6 above)
Comment (falkerson) [reply registry/consul.go]: Sure. So maybe just better to avoid functionality 'CheckKeyValuePairs' in this commit and provide it separately?

Comment (falkerson) [reply registry/consul.go]: but is it necessary to fix the issue? let's keep commits for pull request modular. If this is not a root of problem we should avoid such change in fix commit/PR. And add it in separate. If commit describe that it should be a fix, we provide a fix - nothing more. Also we already have issue for stroing/get config in Consul https://github.com/edgexfoundry/device-sdk-go/issues/6

Comment (weichou1229) [reply registry/consul.go]: Ok, I have refactor GetConsulClient function. Only check consul available and return consulClient.
